File: 2023/01 - Trebuchet‽/main.hs
import Data.Char (isDigit, digitToInt)
import Data.List (isPrefixOf, tails)
import Data.Maybe (mapMaybe)

extractDigits1 :: String -> [Int]
extractDigits1 [] = []
extractDigits1 (x:xs)
    | isDigit x = (digitToInt x) : (extractDigits1 xs)
    | otherwise =                   extractDigits1 xs

extractDigits2 :: String -> [Int]
extractDigits2 str = mapMaybe digit (tails str)
    where
        digit :: String -> Maybe Int
        digit [] = Nothing
        digit str
            | isDigit (head str)       = Just $ digitToInt (head str)
            | "zero"  `isPrefixOf` str = Just 0
            | "one"   `isPrefixOf` str = Just 1
            | "two"   `isPrefixOf` str = Just 2
            | "three" `isPrefixOf` str = Just 3
            | "four"  `isPrefixOf` str = Just 4
            | "five"  `isPrefixOf` str = Just 5
            | "six"   `isPrefixOf` str = Just 6
            | "seven" `isPrefixOf` str = Just 7
            | "eight" `isPrefixOf` str = Just 8
            | "nine"  `isPrefixOf` str = Just 9
            | otherwise                = Nothing

main :: IO ()
main = do
    input <- readFile "input.txt"

    let part1 = sum [ 10*(head nums) + (last nums) | nums <- map extractDigits1 $ lines input ]
    let part2 = sum [ 10*(head nums) + (last nums) | nums <- map extractDigits2 $ lines input ]

    putStr "Part 1: "
    putStrLn $ show $ part1

    putStr "Part 2: "
    putStrLn $ show $ part2

File: 2023/01 - Trebuchet‽/main.lua
local part1 = 0
local part2 = 0

local digits = {"one", "two", "three", "four", "five", "six", "seven", "eight", "nine"} -- no zero!

for line in io.lines("input.txt") do
    local first = string.sub(line, 1, line:find("%d"))
    local last = string.sub(line, #line - line:reverse():find("%d") + 1)

    part1 = part1 + tonumber( first:sub(-1) .. last:sub(1,1) )
    
    local minpos = math.huge
    local num1 = nil
    for i = 1, 9 do
        local pos = first:find(digits[i])
        if pos and pos < minpos then
            minpos = pos
            num1 = i
        end
    end

    local maxpos = -math.huge
    local num2 = nil
    for i = 1, 9 do
        -- TODO: make this cleaner...
        local pos = last:find(digits[i])
        if pos then
            local tmp
            repeat
                tmp = last:find(digits[i], pos+1)
                if tmp then pos = tmp end
            until not tmp

            if pos > maxpos then
                maxpos = pos
                num2 = i
            end
        end
    end

    if not num1 then num1 = tonumber(first:sub(-1)) end
    if not num2 then num2 = tonumber(last:sub(1,1)) end

    part2 = part2 + (10*num1 + num2)
end

io.write("Part 1: ", part1, '\n')
io.write("Part 2: ", part2, '\n')